pkg/vpa/vpa.go

```diff
@@ -162,69 +155,126 @@ func (r Reconciler) namespaceIsManaged(namespace *corev1.Namespace) bool {
 	return r.OnByDefault
 }
 
-func (r Reconciler) reconcileDeploymentsAndVPAs(nsName string, vpaNames, deploymentNames []string, vpaUpdateMode v1beta2.UpdateMode) error {
-	// Create any VPAs that need to be
-	vpaNeeded := utils.Difference(deploymentNames, vpaNames)
-	klog.V(3).Infof("Diff deployments, vpas: %v", vpaNeeded)
+func (r Reconciler) reconcileDeploymentsAndVPAs(nsName string, vpas []v1beta2.VerticalPodAutoscaler, deployments []appsv1.Deployment, vpaUpdateMode v1beta2.UpdateMode) error {
+	// these keys will eventually contain the leftover vpas that do not have a matching deployment associated
+	vpaHasAssociatedDeployment := map[string]bool{}
+	for _, deployment := range deployments {
+		var dvpa *v1beta2.VerticalPodAutoscaler
+		// search for the matching vpa (will have the same name)
+		for _, vpa := range vpas {
+			if deployment.Name == vpa.Name {
+				// found the vpa associated with this deployment
+				dvpa = &vpa
+				vpaHasAssociatedDeployment[dvpa.Name] = true
+				break
+			}
+		}
 
-	if len(vpaNeeded) == 0 {
-		klog.Info("All VPAs are in sync.")
-	} else if len(vpaNeeded) > 0 {
-		for _, vpaName := range vpaNeeded {
-			err := r.createVPA(nsName, vpaName, vpaUpdateMode)
+		// for logging
+		vpaName := "none"
+		if dvpa != nil {
+			vpaName = dvpa.Name
+		}
+		klog.V(2).Infof("Reconciling Namespace/%s for Deployment/%s with VPA/%s in vpa-update-mode=%s", nsName, deployment.Name, vpaName, vpaUpdateMode)
+		err := r.reconcileDeploymentAndVPA(nsName, deployment, dvpa, vpaUpdateMode)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, vpa := range vpas {
+		if _, ok := vpaHasAssociatedDeployment[vpa.Name]; !ok {
```

Comment:
I would suggest just using `if !vpaHasAssociatedDeployment[vpa.Name]`. It will default to `false` if `vpa.Name` isn't present.

That way if the logic above ever changes and we set `vpaHasAssociatedDeployment[vpa.Name] = false` somewhere, this block will still trigger

Reply:
👍 [ ] Do this